domain/port/service/service.go

```diff
@@ -312,3 +261,215 @@ func verifyNoPortRangeConflicts(rangesA, rangesB []network.PortRange) error {
+	for wildcardOpeningPortRange := range wildcardOpeningSet {
+		for _, endpoint := range endpoints {
+			// Leave the wildcard endpoint as is.
+			if endpoint.Endpoint == WildcardEndpoint {
+				continue
+			}
+			closePortsOut[endpoint.Endpoint] = append(closePortsOut[endpoint.Endpoint], wildcardOpeningPortRange)
+		}
+	}
+
+	// Close port ranges closed on the wildcard endpoint on all other endpoints.
+	wildcardClosing, _ := closePortsIn[WildcardEndpoint]
+	for _, wildcardClosingPortRange := range wildcardClosing {
+		// Leave the wildcard endpoint as is.
+		for _, endpoint := range endpoints {
+			if endpoint.Endpoint == WildcardEndpoint {
+				continue
+			}
+			closePortsOut[endpoint.Endpoint] = append(closePortsOut[endpoint.Endpoint], wildcardClosingPortRange)
+		}
+	}
+
+	// construct a map of all port ranges being closed to the endpoint they're
+	// being closed on. Except the wildcard endpoint.
+	closePortsToEndpointMap := make(map[network.PortRange]string)
+	for endpoint, endpointClosePorts := range closePortsOut {
+		for _, portRange := range endpointClosePorts {
+			closePortsToEndpointMap[portRange] = endpoint
+		}
+	}
+	// Ensure endpoints closed on the wildcard endpoint are not in the map.
+	for _, wildcardClosePorts := range closePortsOut[WildcardEndpoint] {
+		delete(closePortsToEndpointMap, wildcardClosePorts)
+	}
+
+	// If we're closing a port range for a specific endpoint which is open
+	// on the wildcard endpoint, we need to close it on the wildcard endpoint
+	// and open it on all other endpoints except the targeted endpoint.
+	for _, portRange := range openedOnWildcard {
+		if endpoint, ok := closePortsToEndpointMap[portRange]; ok {
+
+			// This port range, open on the wildcard endpoint, is being closed
+			// on some endpoint. We need to close it on the wildcard, and open
+			// it on all endpoints other than the wildcard & targeted endpoint.
+			closePortsOut[WildcardEndpoint] = append(closePortsOut[WildcardEndpoint], portRange)
+
+			for _, otherEndpoint := range endpoints {
+				if otherEndpoint.Endpoint == WildcardEndpoint || otherEndpoint.Endpoint == endpoint {
+					continue
+				}
+				openPortsOut[otherEndpoint.Endpoint] = append(openPortsOut[otherEndpoint.Endpoint], portRange)
+			}
+
+			// Remove the port range from openPorts for the targeted endpoint.
+			for i, otherPortRange := range openPortsOut[endpoint] {
+				if otherPortRange == portRange {
+					openPortsOut[endpoint] = append(openPortsOut[endpoint][:i], openPortsOut[endpoint][i+1:]...)
+					break
+				}
+			}
+		}
+	}
+
+	return openPortsOut, closePortsOut, nil
+}
+
+func filterOutAlreadyOpenRanges(current portRangeUUIDIndex, toAdd network.GroupedPortRanges) network.GroupedPortRanges {
+	filtered := network.GroupedPortRanges{}
+	for endpoint, portRanges := range toAdd {
+		for _, portRange := range portRanges {
+			if _, ok := current[endpoint][portRange]; !ok {
+				filtered[endpoint] = append(filtered[endpoint], portRange)
+			}
+		}
+	}
```

Comment:
Is it possible to skip the inner for loop by checking if `current[endpoint]` is not there?